lib/src/lifecycle_module.dart

```diff
@@ -207,45 +266,78 @@ abstract class LifecycleModule extends Object with Disposable {
   /// Suspend indicates to the module that it should go into a low-activity
   /// state. For example, by disconnecting from backend services and unloading
   /// heavy data structures.
-  Future suspend() {
-    Completer completer = new Completer();
-    if (!_isSuspended && _isLoaded) {
-      _willSuspendController.add(this);
-      Future.wait(_childModules.map((c) => c.suspend())).then((_) {
-        onSuspend().then((_) {
-          _didSuspendController.add(this);
-          _isSuspended = true;
-          completer.complete();
-        });
-      });
-    } else {
-      _logger.warning(
-          'Module "$name" is ${_isLoaded ? 'already suspended' : 'not loaded'}, cannot suspend.');
-      completer.complete();
+  ///
+  /// Initiates the suspend process when the module is in the loaded state. If
+  /// the module is in the suspended or suspending state a warning is logged and
+  /// the method is a noop. If the module is in any other state, a StateError is
+  /// thrown.
+  Future<Null> suspend() {
+    if (isSuspending) {
+      _logger.warning('Module "$name" is already suspending, cannot suspend.');
+      return _transitionFuture;
+    }
+
+    if (isSuspended) {
+      _logger.warning('Module "$name" is already suspended, cannot suspend.');
+      return new Future.value(null);
     }
+
+    if (!isLoaded) {
+      return new Future.error(
+          new StateError('Module "$name" state is $_state, cannot suspend.'));
+    }
+
+    _state = LifecycleState.suspending;
+    final completer = new Completer<Null>();
+    _transitionFuture = completer.future;
+    _willSuspendController.add(this);
+
+    Future.wait(_childModules.map((c) => c.suspend())).then((_) {
+      onSuspend().then((_) {
+        _didSuspendController.add(this);
+        _state = LifecycleState.suspended;
+        _transitionFuture = null;
+        completer.complete();
+      });
+    });
+
     return completer.future;
   }
 
   /// Public method to resume the module.
   ///
   /// This should put the module back into its normal state after the module
   /// was suspended.
-  Future resume() {
-    Completer completer = new Completer();
-    if (_isSuspended && _isLoaded) {
-      _willResumeController.add(this);
-      Future.wait(_childModules.map((c) => c.resume())).then((_) {
-        onResume().then((_) {
-          _didResumeController.add(this);
-          _isSuspended = false;
-          completer.complete();
-        });
-      });
-    } else {
-      _logger.warning(
-          'Module "$name" is ${_isLoaded ? 'not suspended' : 'not loaded'}, cannot suspend.');
-      completer.complete();
+  ///
+  /// Only initiates the resume process when the module is in the suspended
+  /// state. If the module is in the resuming state a warning is logged and the
```

Comment:
@todbachman-wf and I talked about the legality of a state transition and reasoned that a `StateError` is correct whenever a consumer attempts to change states without following a green line in the PR diagram. I guess it is assumed that there is an implicit self referential transition for every state. Your correct in pointing out this is the only update that does not follow this pattern.

I would argue that this should be updated to **NOT** throw **AND** noop  when already in a loaded state.